### PR TITLE
network: Allow to set static hostname

### PR DESCRIPTION
**Pull request by ssahani**
closes https://github.com/systemd/systemd/issues/17403

### CODE REVIEW

File: man/systemd.network.xml [hunk 1]
```diff
@@ -1645,6 +1645,14 @@ IPv6Token=prefixstable:2002:da8:1::</programlisting></para>
             no spaces or dots, and be formatted as a valid DNS domain name.</para>
           </listitem>
         </varlistentry>
+        <varlistentry>
+          <term><varname>SetStaticHostname=</varname></term>
```
Comment (yuwata) [inline man/systemd.network.xml]: I think it is better to not introduce a new setting, but accept `UseHostname=static`.

File: src/network/networkd-network.h [hunk 2]
```diff
@@ -135,6 +135,7 @@ struct Network {
         int dhcp_use_gateway;
         bool dhcp_use_timezone;
         bool dhcp_use_hostname;
+        bool dhcp_set_static_hostname;
```
Comment (yuwata) [inline src/network/networkd-network.h]: Maybe, its better to introduce a enum:
```
typedef enum UseHostname {
    USE_HOSTNAME_NO,
    USE_HOSTNAME_YES,
    USE_HOSTNAME_STATIC,
    _USE_HOSTNAME_MAX,
    _USE_HOSTNAM_INVALID = -1,
} UseHostname;
```
and then introduce a `_from_string()` with `DEFINE_PRIVATE_STRING_TABLE_LOOKUP_FROM_STRING_WITH_BOOLEAN()`.

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -949,6 +955,12 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
                         if (r < 0)
                                 log_link_error_errno(link, r, "Failed to set transient hostname to '%s': %m", hostname);
                 }
+
+                if (hostname && link->network->dhcp_set_static_hostname) {
+                        r = manager_set_static_hostname(link->manager, hostname);
+                        if (r < 0)
+                                log_link_error_errno(link, r, "Failed to set static hostname to '%s': %m", hostname);
+                }
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: Is it necessary to also set the 'transient' hostname in the above? I do not think so...

File: src/network/networkd-manager.c [hunk 4]
```diff
@@ -1204,6 +1204,52 @@ int manager_set_hostname(Manager *m, const char *hostname) {
...
+        e = sd_bus_message_get_error(m);
+        if (e) {
+                r = sd_bus_error_get_errno(e);
+                log_warning_errno(r, "Could not set static hostname: %s", bus_error_message(e, r));
+        }
+
+        return 1;
+}
+
+int manager_set_static_hostname(Manager *m, const char *hostname) {
```
Comment (yuwata) [inline src/network/networkd-manager.c]: How about to merge this with `manager_set_hostname()`? They are almost equivalent.

File: src/network/networkd-dhcp4.c [hunk 5]
```diff
@@ -465,6 +465,12 @@ static int dhcp_reset_hostname(Link *link) {
         if (r < 0)
                 return log_link_error_errno(link, r, "DHCP error: Failed to reset transient hostname: %m");
 
+        if (link->network->dhcp_set_static_hostname) {
+                r = manager_set_static_hostname(link->manager, NULL);
+                if (r < 0)
+                        return log_link_error_errno(link, r, "DHCP error: Failed to reset static hostname: %m");
+        }
+
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c]: I do not think we should unset the static hostname. If we do so, then every time we restart or stop networkd, the static hostname is unset. We already unset the transient hostname in the above, so it makes the system's hostname become `FALLBACK_HOSTNAME`, especially every time on shutdown the system...

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ssahani** [comment]
Updated thanks for the review. ptal

**Comment by bluca** [comment]
CI looks unhappy:

```
582/582 check-directives                        FAIL     0.07 s (exit status 1)

--- command ---
13:34:36 /home/runner/work/systemd/systemd/tools/check-directives.sh /home/runner/work/systemd/systemd
--- stdout ---
378a379
> SetStaticHostname=
Looks like test/fuzz/fuzz-network-parser/directives.network hasn't been updated
```

**Comment by ssahani** [comment]
> CI looks unhappy:
> 
> ```
> 582/582 check-directives                        FAIL     0.07 s (exit status 1)
> 
> --- command ---
> 13:34:36 /home/runner/work/systemd/systemd/tools/check-directives.sh /home/runner/work/systemd/systemd
> --- stdout ---
> 378a379
> > SetStaticHostname=
> Looks like test/fuzz/fuzz-network-parser/directives.network hasn't been updated
> ```

Yes dropped now . thanks !

File: src/network/networkd-dhcp4.c [hunk 6]
```diff
@@ -22,6 +22,15 @@
 #include "sysctl-util.h"
 #include "web-util.h"
 
+static const char *const use_hostname_table[_USE_HOSTNAME_MAX] = {
+        [USE_HOSTNAME_NO]     = "no",
+        [USE_HOSTNAME_YES]    = "yes",
+        [USE_HOSTNAME_STATIC] = "static",
+        [USE_HOSTNAME_BOTH]   = "both",
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c:29]: `both`? I do not have any reason to set both transient and static hostname. Note that `SetStaticHostname` method updates /etc/hostname and sets kernel hostname. While `SetHostname` just set the kernel hostname. So, calling both method is meaningless.

File: src/network/networkd-dhcp4.c [hunk 7]
```diff
@@ -928,8 +937,8 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
         }
 
         if (link->network->dhcp_use_hostname) {
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c:939]: `if (IN_SET(link->network->dhcp_use_hostname, USE_HOSTNAME_YES, USE_HOSTNMAE_STATIC))`

File: src/network/networkd-dhcp4.c [hunk 8]
```diff
@@ -944,11 +953,19 @@ static int dhcp_lease_acquired(sd_dhcp_client *client, Link *link) {
...
+                        r = manager_set_hostname(link->manager, hostname, false);
                         if (r < 0)
                                 log_link_error_errno(link, r, "Failed to set transient hostname to '%s': %m", hostname);
                 }
+
+                if (hostname && IN_SET(link->network->dhcp_use_hostname, USE_HOSTNAME_STATIC, USE_HOSTNAME_BOTH)) {
+                        r = manager_set_hostname(link->manager, hostname, true);
+                        if (r < 0)
+                                log_link_error_errno(link, r, "Failed to set static hostname to '%s': %m", hostname);
+                }
```
Comment (yuwata) [inline src/network/networkd-dhcp4.c:968]: Please move the debug log in `if (hostname)` block. Also, as already commented in the above, `USE_HOSTNAME_BOTH` is meaningless, so how about something like the following:
```
if (hostname) {
    log_link_debug(link, "Setting...");
    r = manager_set_hostname(link->manager, hostname, link->network->dhcp_use_hostname == USE_HOSTNAME_STATIC);
    if (r < 0)
         log_link_error_errno(...);
}
```

File: src/network/networkd-manager.c [hunk 9]
```diff
@@ -98,7 +98,7 @@ static int on_connected(sd_bus_message *message, void *userdata, sd_bus_error *r
 
         /* Did we get a timezone or transient hostname from DHCP while D-Bus wasn't up yet? */
         if (m->dynamic_hostname)
-                (void) manager_set_hostname(m, m->dynamic_hostname);
+                (void) manager_set_hostname(m, m->dynamic_hostname, false);
```
Comment (yuwata) [inline src/network/networkd-manager.c:101]: If `bus` is not ready when setting hostname at first, then, this always try to set transient hostname...

How about the following?
```
# networkd-manager.h
Manager {
    (...);
    char *transient_hostname;
    char *static_hostname;
    (...);
}

# networkd-manager.c
if (m->static_hostname)
    (void) manager_set_hostname(m, m->static_hostname, true);
else if (m->transient_hostname)
    (void) manager_set_hostname(m, m->transient_hostname, false);
```

File: src/network/networkd-manager.c [hunk 10]
```diff
@@ -1172,34 +1172,44 @@ static int set_hostname_handler(sd_bus_message *m, void *userdata, sd_bus_error
...
         int r;
 
-        log_debug("Setting transient hostname: '%s'", strna(hostname));
+        log_debug("Setting %s hostname: '%s'", static_hostname ? "static" : "transient", strna(hostname));
 
-        if (free_and_strdup(&m->dynamic_hostname, hostname) < 0)
-                return log_oom();
+        if (!static_hostname)
+                if (free_and_strdup(&m->dynamic_hostname, hostname) < 0)
+                        return log_oom();
```
Comment (yuwata) [inline src/network/networkd-manager.c:1183]: Then, 
```
if (static_hostname)
    r = free_and_strdup(&m->static_hostname, hostname);
else
    r = free_and_strdup(&m->transient_hostname, hostname);
if (r < 0)
    return log_oom();
```

File: src/network/networkd-manager.c [hunk 11]
```diff
@@ -1172,34 +1172,44 @@ static int set_hostname_handler(sd_bus_message *m, void *userdata, sd_bus_error
...
         }
 
+        if (static_hostname)
+                kind = strdup("SetHostname");
+        else
+                kind = strdup("SetStaticHostname");
+
+        if (!kind)
+                return log_oom();
+
```
Comment (yuwata) [inline src/network/networkd-manager.c:1197]: It is not necessary to `strdup()` the method name.

File: src/network/networkd-manager.c [hunk 12]
```diff
@@ -1172,34 +1172,44 @@ static int set_hostname_handler(sd_bus_message *m, void *userdata, sd_bus_error
...
         r = sd_bus_call_method_async(
                         m->bus,
                         NULL,
                         "org.freedesktop.hostname1",
                         "/org/freedesktop/hostname1",
                         "org.freedesktop.hostname1",
-                        "SetHostname",
+                        kind,
```
Comment (yuwata) [inline src/network/networkd-manager.c:1204]: Simply, `static_hostname ? "SetStaticHostname" : "SetHostname"` is enough.

File: src/network/networkd-manager.h [hunk 13]
```diff
@@ -96,7 +96,7 @@ void manager_dirty(Manager *m);
 
 Link* manager_find_uplink(Manager *m, Link *exclude);
 
-int manager_set_hostname(Manager *m, const char *hostname);
+int manager_set_hostname(Manager *m, const char *hostname, bool static_hostname);
```
Comment (yuwata) [inline src/network/networkd-manager.h:99]: Not a big matter, but the name `bool static_hostname` is redundant, `bool static` or `bool is_static` is enough.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/network/networkd-dhcp4.c (hunk 6 above)
Comment (ssahani) [reply src/network/networkd-dhcp4.c:29]: The idea behind the both is clear logic what going to happen as you said above. If you think it's reductact we can drop this and update in man only.

**Pull request closed without merging**